### PR TITLE
✨ Implement ImmutableSegmentedHashSet<T>

### PR DESCRIPTION
**Pull request by sharwell**
Builds on #54574

### CODE REVIEW

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler this is now ready for review

File: src/Dependencies/Collections/ImmutableSegmentedHashSet`1.cs [hunk 1]
```diff
@@ -12,148 +12,277 @@ namespace Microsoft.CodeAnalysis.Collections
...
+                var set = new SegmentedHashSet<T>(self.KeyComparer) { value };
+                return new ImmutableSegmentedHashSet<T>(set);
+            }
+            else if (self.Contains(value))
+            {
+                return self;
+            }
+            else
+            {
+                // TODO: Avoid the builder allocation
```
Comment (333fred) [inline src/Dependencies/Collections/ImmutableSegmentedHashSet`1.cs:70]: Are these todo's from the original source?

Comment (sharwell) [reply src/Dependencies/Collections/ImmutableSegmentedHashSet`1.cs:70]: No, these are ways the current implementation could theoretically be improved. My hope is that if someone sees allocations of certain types in the future, they are directed to the most straightforward way to eliminate those allocations.

Comment (sharwell) [reply src/Dependencies/Collections/ImmutableSegmentedHashSet`1.cs:70]: Here's an example of how I broke out a `ValueBuilder` nested type to reduce builder allocations in `ImmutableSegmentedList<T>`:
https://github.com/dotnet/roslyn/blob/main/src/Dependencies/Collections/ImmutableSegmentedList%601%2BValueBuilder.cs

**Review by 333fred — APPROVED** [review]

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for second review

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for second review

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for second review

**Comment by sharwell** [comment]
https://github.com/orgs/dotnet/teams/roslyn-compiler for second review

**Comment by sharwell** [comment]
@dotnet/roslyn-compiler for second review

**Comment by jaredpar** [comment]
How can we see the delta from what the BCL sources were to what you added as a part of moving to Roslyn? I'm less interested in reviewing the BCL sources (battle tested, already reviewed, etc ...). Think it's more interesting for us to look at the delta to see what has changed and evaluate that.

**Comment by sharwell** [comment]
@jaredpar There is no exact mirror for this type in the BCL. The closest we have is the AVL-backed `ImmutableHashSet<T>`. In this case, the sources copied from runtime are the _test_ cases for `ImmutableHashSet<T>` which were updated in 03c2ae73a22c6fc3f74ee5ae9f7f9be7ede00ee1 to execute against the new data type and followed by a2023ec2ee0f5425b8da2ceea77525fb0642be6a to get the tests fully passing against the new structure.

**Review by jaredpar — APPROVED** [review]

File: src/Dependencies/Collections/ImmutableSegmentedHashSet`1.cs [hunk 2]
```diff
@@ -0,0 +1,344 @@
...
+                var builder = self.ToBuilder();
+                builder.Add(value);
+                return builder.ToImmutable();
+            }
+        }
+
+        /// <inheritdoc cref="ImmutableHashSet{T}.Clear()"/>
+        public ImmutableSegmentedHashSet<T> Clear()
+        {
+            var self = this;
```
[the remark this answers is not in the document]
Comment (sharwell) [reply src/Dependencies/Collections/ImmutableSegmentedHashSet`1.cs:81]: Good catch. This type appears to be missing the equivalent of this notice which is applied to `ImmutableSegmentedList<T>`:

https://github.com/dotnet/roslyn/blob/519058e79ef19f1dc56cc86b65b103e80523799f/src/Dependencies/Collections/ImmutableSegmentedList%601.cs#L59-L69

[the remark this answers is not in the document]
Comment (sharwell) [reply src/Dependencies/Collections/ImmutableSegmentedHashSet`1.cs:81]: I'll file a follow-up issue to add this documentation and implement the value builder.

**Pull request merged**